feat: implement OneKey ID authentication with OAuth support

sidmorizon opened the pull request:
## <!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Implemented OAuth authentication with support for Google and Apple sign-in across multiple platforms (web, desktop, and extension).
  * Added persistent sessions with automatic token refresh capabilities for seamless user experience.
  * Introduced email-based OTP authentication as an alternative sign-in method.
  * Enhanced authentication and session management infrastructure.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->
Implement OneKey ID authentication service with OAuth support for Google and Apple sign-in across all platforms (desktop, mobile, web, and browser extension).

## Key Changes

### ServiceOneKeyID
- Added new `ServiceOneKeyID` for managing OneKey ID authentication
- Integrated login with access token, logout, and login status checking

### OAuth Implementation
- Enhanced `useSupabaseAuth` to support OAuth sign-in methods for Google and Apple
- Implemented platform-specific OAuth popup handling:
  - Desktop: Support for WEBVIEW and DEEP_LINK methods
  - Extension: Added OAuth callback page with identity permissions
  - Web: Enhanced popup handling with session management
  - Native: Platform-specific redirect handling

### Session Management
- Introduced temporary Supabase client for OAuth sessions without automatic persistence
- Added optional session persistence options for better flexibility
- Implemented polling mechanism to keep OAuth popup focused during authentication

### Developer Tools
- Added `OneKeyIDGallery` component for testing authentication features
- Included JWT decoding utility for access token handling

## Files Modified
- `packages/kit-bg/src/services/ServiceOneKeyID/` - New authentication service
- `packages/kit/src/components/OneKeyAuth/` - OAuth popup handlers for all platforms
- `packages/kit/src/components/OneKeyAuth/supabase/` - Supabase auth integration
- `apps/ext/` - Extension manifest and OAuth callback page
- `packages/shared/src/utils/stringUtils.ts` - JWT decoding utility

sidmorizon committed:
feat: implement OneKey ID authentication service and integrate OAuth methods

- Added ServiceOneKeyID for managing OneKey ID authentication, including login with access token, logout, and checking login status.
- Enhanced useSupabaseAuth to support OAuth sign-in methods for Google and Apple.
- Introduced OneKeyIDGallery component for testing OneKey ID authentication features.
- Updated BackgroundApi and BackgroundApiProxy to include the new OneKey ID service.
- Added new routes and components to the Developer Gallery for showcasing OneKey ID functionality.

sidmorizon committed:
feat: enhance OAuth handling in useSupabaseAuth for desktop

- Introduced support for two OAuth methods in the desktop environment: WEBVIEW and DEEP_LINK.
- Updated redirect logic to accommodate both methods, improving user experience and reliability.
- Added detailed comments and documentation for each method's pros and cons.
- Enhanced the UI for the OAuth webview, including a close button and error handling for navigation events.
- Implemented timeout handling for OAuth sign-in processes to improve robustness.

sidmorizon committed:
feat: implement OAuth callback handling and enhance extension support

- Added a new HTML file for OAuth callback handling in Chrome extensions.
- Updated the manifest files to include necessary permissions and OAuth2 configuration.
- Enhanced the OAuth handling logic in the extension to support identity permissions.
- Improved the overall structure and organization of OAuth-related functions across platforms.
- Updated the webpack configuration to include the new OAuth callback page.
- Refactored the useSupabaseAuth hook to streamline OAuth processes for different platforms.

sidmorizon committed:
feat: enhance OAuth session handling and refactor related components

- Introduced a temporary Supabase client to manage OAuth sessions without automatic persistence.
- Updated various OAuth functions to accept options for session persistence, improving flexibility.
- Refactored the `useSupabaseAuth` hook to streamline OAuth processes across platforms.
- Added detailed logging for OAuth session handling to aid in debugging.
- Enhanced the UI components to include session persistence options for better user control.

sidmorizon committed:
feat: enhance OAuth popup handling and session management

- Added a polling mechanism to keep the OAuth popup window focused during authentication, improving user experience.
- Refactored the `openOAuthPopupExtIdentity` function to manage window focus and cleanup more effectively.
- Updated `openOAuthPopupWeb` to allow optional session persistence in the OAuth session parameters.
- Enhanced `useSupabaseAuth` to support optional session persistence for better flexibility.
- Introduced JWT decoding utility in `stringUtils` for improved access token handling in the UI components.

revan-zhang commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/84c4e5e2-d101-468d-9d8c-993c842dafc3?refs=2afae75bada33e3d4aba5e337f9c5b348ec632de&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/84c4e5e2-d101-468d-9d8c-993c842dafc3/license?refs=2afae75bada33e3d4aba5e337f9c5b348ec632de&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

sidmorizon committed:
chore: update .gitignore and remove debug log file

- Added .cursor/debug.log to .gitignore to prevent tracking of debug logs.
- Deleted the .cursor/debug.log file to clean up unnecessary debug information.

sidmorizon commented on apps/ext/src/manifest/chrome_v3.js:
```diff
@@ -123,4 +124,17 @@ module.exports = {
+  // OAuth2 configuration for chrome.identity.getAuthToken
+  // Required for CHROME_GET_AUTH_TOKEN method
+  // The client_id should be a Chrome Extension type OAuth client from Google Cloud Console
+  'oauth2': {
```
remove

limichange committed:
Merge branch 'x' into fix/keyless-google-login

limichange commented:
@coderabbitai resolve
